tiledb/api/src/range.rs

```diff
@@ -0,0 +1,841 @@
+                }
+            }
+        )+
+    }
+}
+
+var_value_range_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+var_value_range_from!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+var_value_range_from!(Float32: f32, Float64: f64);
+
+impl From<(&str, &str)> for VarValueRange {
+    fn from(value: (&str, &str)) -> VarValueRange {
+        let start = value.0.as_bytes().to_vec().into_boxed_slice();
+        let end = value.1.as_bytes().to_vec().into_boxed_slice();
+        VarValueRange::UInt8(start, end)
+    }
+}
+
+impl From<&[&str; 2]> for VarValueRange {
+    fn from(value: &[&str; 2]) -> VarValueRange {
+        let start = value[0].as_bytes().to_vec().into_boxed_slice();
+        let end = value[1].as_bytes().to_vec().into_boxed_slice();
+        VarValueRange::UInt8(start, end)
+    }
+}
+
+#[macro_export]
+macro_rules! var_value_range_go {
+    ($expr:expr, $DT:ident, $start:pat, $end:pat, $then:expr) => {
+        match $expr {
+            VarValueRange::UInt8($start, $end) => {
+                type $DT = u8;
+                $then
+            }
+            VarValueRange::UInt16($start, $end) => {
+                type $DT = u16;
+                $then
+            }
+            VarValueRange::UInt32($start, $end) => {
+                type $DT = u32;
+                $then
+            }
+            VarValueRange::UInt64($start, $end) => {
+                type $DT = u64;
+                $then
+            }
+            VarValueRange::Int8($start, $end) => {
+                type $DT = i8;
+                $then
+            }
+            VarValueRange::Int16($start, $end) => {
+                type $DT = i16;
+                $then
+            }
+            VarValueRange::Int32($start, $end) => {
+                type $DT = i32;
+                $then
+            }
+            VarValueRange::Int64($start, $end) => {
+                type $DT = i64;
+                $then
+            }
+            VarValueRange::Float32($start, $end) => {
+                type $DT = f32;
+                $then
+            }
+            VarValueRange::Float64($start, $end) => {
+                type $DT = f64;
+                $then
+            }
+        }
+    };
+}
+
+#[derive(Clone, Deserialize, Serialize, PartialEq)]
+pub enum Range {
+    Single(SingleValueRange),
+    Multi(MultiValueRange),
+    Var(VarValueRange),
+}
+
+impl Range {
+    // N.B. This is not `check_field_compatibility` because dimensions have
+    // restrictions on their cell_val_num that don't apply to attributes.
+    //
+    // See: tiledb::sm::Dimension::set_cell_val_num
+    pub fn check_dimension_compatibility(
+        &self,
+        datatype: Datatype,
+        cell_val_num: CellValNum,
+    ) -> TileDBResult<()> {
+        match self {
+            Self::Single(svr) => svr.check_datatype(datatype)?,
+            Self::Multi(_) => {
+                return Err(Error::InvalidArgument(anyhow!(
```

Comment:
This surprises me a bit, it seems like it would be just easier to support than Var. Oh well.

Reply:
AFAIK, its not that it'd be difficult to implement, but that it opens up a bunch of worm cans with respect to the semantics and what not.